x-pack/plugins/maps/public/classes/sources/table_source/table_source.ts

```diff
@@ -0,0 +1,218 @@
+ */
+
+import uuid from 'uuid';
+import { FIELD_ORIGIN, SOURCE_TYPES, VECTOR_SHAPE_TYPE } from '../../../../common/constants';
+import {
+  MapExtent,
+  MapFilters,
+  MapQuery,
+  TableSourceDescriptor,
+  VectorJoinSourceRequestMeta,
+  VectorSourceSyncMeta,
+} from '../../../../common/descriptor_types';
+import { Adapters } from '../../../../../../../src/plugins/inspector/common/adapters';
+import { ITermJoinSource } from '../term_join_source';
+import { BucketProperties, PropertiesMap } from '../../../../common/elasticsearch_util';
+import { IField } from '../../fields/field';
+import { Query } from '../../../../../../../src/plugins/data/common/query';
+import {
+  AbstractVectorSource,
+  BoundsFilters,
+  GeoJsonWithMeta,
+  IVectorSource,
+  SourceTooltipConfig,
+} from '../vector_source';
+import { DataRequest } from '../../util/data_request';
+import { InlineField } from '../../fields/inline_field';
+
+export class TableSource extends AbstractVectorSource implements ITermJoinSource, IVectorSource {
+  static type = SOURCE_TYPES.TABLE_SOURCE;
+
+  static createDescriptor(descriptor: Partial<TableSourceDescriptor>): TableSourceDescriptor {
+    return {
+      type: SOURCE_TYPES.TABLE_SOURCE,
+      __rows: descriptor.__rows || [],
+      __columns: descriptor.__columns || [],
+      term: descriptor.term || '',
+      id: descriptor.id || uuid(),
+    };
+  }
+
+  readonly _descriptor: TableSourceDescriptor;
+
+  constructor(descriptor: Partial<TableSourceDescriptor>, inspectorAdapters?: Adapters) {
+    const sourceDescriptor = TableSource.createDescriptor(descriptor);
+    super(sourceDescriptor, inspectorAdapters);
+    this._descriptor = sourceDescriptor;
+  }
+
+  async getDisplayName(): Promise<string> {
+    // no need to localize. this is never rendered.
+    return `table source ${uuid()}`;
+  }
+
+  getSyncMeta(): VectorSourceSyncMeta | null {
+    return null;
+  }
+
+  async getPropertiesMap(
+    searchFilters: VectorJoinSourceRequestMeta,
+    leftSourceName: string,
+    leftFieldName: string,
+    registerCancelCallback: (callback: () => void) => void
+  ): Promise<PropertiesMap> {
+    const propertiesMap: PropertiesMap = new Map<string, BucketProperties>();
+
+    const fieldNames = await this.getFieldNames();
+
+    for (let i = 0; i < this._descriptor.__rows.length; i++) {
+      const row: { [key: string]: string | number } = this._descriptor.__rows[i];
+      let propKey: string | number | undefined;
+      const props: { [key: string]: string | number } = {};
+      for (const key in row) {
+        if (row.hasOwnProperty(key)) {
+          if (key === this._descriptor.term && row[key]) {
+            propKey = row[key];
+          }
+          if (fieldNames.indexOf(key) >= 0 && key !== this._descriptor.term) {
+            props[key] = row[key];
+          }
+        }
+      }
+      if (propKey && !propertiesMap.has(propKey.toString())) {
```

Comment:
Not sure `propKey.toString()` is needed since key is typed as string. To be consistent with ESTermSource, should key be typed as `string | number`?

Reply:
will keep as is. `propKey` can be `string | number` (e.g. `40206`, `50103` for a `key` that must be a `string`). so we still need to coerce to put it in the PropertiesMap.